source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,300 @@
+    }
+  }
+
+protected:
+  // The parent group verifier.
+  const BaseVerifierImpl* const parent_;
+};
+
+// Provider specific verifier
+class ProviderVerifierImpl : public BaseVerifierImpl {
+public:
+  ProviderVerifierImpl(const std::string& provider_name, const AuthFactory& factory,
+                       const JwtProvider& provider, const BaseVerifierImpl* parent)
+      : BaseVerifierImpl(parent), auth_factory_(factory), extractor_(Extractor::create(provider)),
+        provider_name_(absl::optional<std::string>{provider_name}) {}
+
+  void verify(ContextSharedPtr context) const override {
+    auto& ctximpl = static_cast<ContextImpl&>(*context);
+    auto auth = auth_factory_.create(getSuppiler(), provider_name_);
+    extractor_->sanitizePayloadHeaders(ctximpl.headers());
+    auth->verify(
+        ctximpl.headers(), extractor_->extract(ctximpl.headers()),
+        [=](const Status& status) { onComplete(status, static_cast<ContextImpl&>(*context)); });
+    if (!ctximpl.getCompletionState(this).is_completed_) {
+      ctximpl.storeAuth(std::move(auth));
+    }
+  }
+
+protected:
+  virtual const CheckAudience* getSuppiler() const { return nullptr; }
+
+private:
+  const AuthFactory& auth_factory_;
+  const ExtractorConstPtr extractor_;
+  const absl::optional<std::string> provider_name_;
+};
+
+class ProviderAndAudienceVerifierImpl : public ProviderVerifierImpl {
+public:
+  ProviderAndAudienceVerifierImpl(const std::string& provider_name, const AuthFactory& factory,
+                                  const JwtProvider& provider, const BaseVerifierImpl* parent,
+                                  const std::vector<std::string>& config_audiences)
+      : ProviderVerifierImpl(provider_name, factory, provider, parent),
+        check_audience_(std::make_unique<CheckAudience>(config_audiences)) {}
+
+private:
+  const CheckAudience* getSuppiler() const override { return check_audience_.get(); }
+
+  // Check audience object
+  ::google::jwt_verify::CheckAudiencePtr check_audience_;
+};
+
+// Allow missing or failed verifier
+class AllowFailedVerifierImpl : public BaseVerifierImpl {
+public:
+  AllowFailedVerifierImpl(const AuthFactory& factory, const Extractor& extractor,
+                          const BaseVerifierImpl* parent)
+      : BaseVerifierImpl(parent), auth_factory_(factory), extractor_(extractor) {}
+
+  void verify(ContextSharedPtr context) const override {
+    auto& ctximpl = static_cast<ContextImpl&>(*context);
+    auto auth = auth_factory_.create({}, absl::nullopt);
+    extractor_.sanitizePayloadHeaders(ctximpl.headers());
+    auth->verify(
+        ctximpl.headers(), extractor_.extract(ctximpl.headers()),
+        [=](const Status& status) { onComplete(status, static_cast<ContextImpl&>(*context)); });
+    if (!ctximpl.getCompletionState(this).is_completed_) {
+      ctximpl.storeAuth(std::move(auth));
+    }
+  }
+
+private:
+  const AuthFactory& auth_factory_;
+  const Extractor& extractor_;
```

Comment:
name this as default_extractor_   with comment as default extractor is from all configured providers.

Reply:
ok then. not need to rename the one in this class